packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/filters.rs

```diff
@@ -0,0 +1,113 @@
+//! Abstract operations that can be used to build up complex queries
+
+use crate::graph_node::GraphNode;
+use std::fmt::Debug;
+
+/// Trait that lets you test to filter different values
+pub trait ApplyTest<TestableValue, Dependency>
+where
+    Self: Debug,
+{
+    /// Test `value` against this filter
+    fn apply_test(&self, value: &TestableValue) -> bool;
+
+    /// Similar to `apply_test` except it returns every "dependency" that was used
+    /// to evaluate the test. This is used to accumulate props referenced when evaluating
+    /// a filter.
+    #[allow(unused_variables)]
+    fn accumulate_deps(&self, value: &TestableValue) -> Vec<Dependency>;
+    //{
+    //    unimplemented!("accumulate_deps not implemented for this filter")
+    // }
+}
+
+/// Operations used to build up complex queries
+#[derive(Debug, Clone, PartialEq)]
+pub enum Op<Left, Right> {
+    /// Satisfied if both `Left` and `Right` are satisfied
+    And(Left, Right),
+    /// Satisfied if either `Left` or `Right` are satisfied
+    Or(Left, Right),
+}
+
+/// Not operation used to build complex queries
+/// Satisfied only if `T` is not satisfied
+#[derive(Debug, Clone, PartialEq)]
+pub struct OpNot<T>(pub T);
+
+impl<T, Left, Right> ApplyTest<T, GraphNode> for Op<Left, Right>
```

Comment:
It's confusing how `Op` is implemented with `GraphNode` and `OpNot` is implemented with generic `D`. You run into trouble with the generic `D` only for cases where the return type of `accumulate_deps` cannot be inferred.

My suggestion is to make `Op` and `OpNot` be implemented with generic `D`, since they can always be inferred from their arguments. Then, make `bool` (as well as `Cond`) be implemented only for `GraphNode`. Those are the two troublesome implementations as they don't have dependencies. (And it's more pleasing to have `Op` and `OpNot` symmetric, especially if we eventually want to combine them.)